File: Java/dynamic_array/DynamicArray.java
package dynamic_array;

import java.util.Arrays;

public class DynamicArray{
    private int length;
    private int capacity;
    private double scalingFactor;
    private int[] array;


    // Constructors
    public DynamicArray(){
        this.length = 0;
        this.capacity = 1;
        this.scalingFactor = 2;
        this.array = new int[this.capacity];
    }

    public DynamicArray(double scalingFactor, int capacity){
        this.length = 0;
        this.capacity = capacity;
        this.scalingFactor = scalingFactor;
        this.array = new int[this.capacity];
    }

    public DynamicArray(double scalingFactor, int length, int defaultValue){
        this.length = length;
        this.capacity = length;
        this.scalingFactor = scalingFactor;
        this.array = new int[this.capacity];
        Arrays.fill(this.array, defaultValue);
    }

    public DynamicArray(int[] array){
        this.length = array.length;
        this.capacity = array.length;
        this.scalingFactor = 2;
        this.array = array;
    }

    public DynamicArray(int[] array, double scalingFactor){
        this.length = array.length;
        this.capacity = array.length;
        this.scalingFactor = scalingFactor;
        this.array = array;
    }

    // Getters
    int getLength(){
        return this.length;
    }

    int getCapacity(){
        return this.capacity;
    }

    double getScalingFactor(){
        return this.scalingFactor;
    }

    int[] getArray(){
        return this.array;
    }

    // Setters
    void setLength(int length){
        this.length = length;
    }

    void setCapacity(int capacity){
        this.capacity = capacity;
    }

    void setScalingFactor(double scalingFactor){
        this.scalingFactor = scalingFactor;
    }

    // Methods
    void resize(){
        this.capacity *= this.scalingFactor;
        int[] newArray = new int[this.capacity];
        for (int i = 0; i < this.length; i++) {
            newArray[i] = this.array[i];
        }
        this.array = newArray;
    }

    void append(int value){
        if (this.length == this.capacity){
            this.resize();
        }
        this.array[this.length] = value;
        this.length++;
    }

    void prepend(int value){
        if(this.length == this.capacity){
            this.resize();
        }

        for(int i = 0; i < this.length; i++){
            this.array[i+1] = this.array[i];
        }
        this.array[0] = value;
        this.length++;
    }

    void insert (int value, int idx) {
        if (this.length == this.capacity){
            this.resize();
        }

        if (idx < 1){
            this.prepend(value);
            return;
        }
        else if (idx >= this.length){
            this.append(value);
            return;
        }
        
        for(int i = idx; i < this.length; i++){
            this.array[i+1] = this.array[i];
        }
        this.array[idx] = value;
    }

    int findFirst(int data){
        int idx = -1;
        for (int i = 0; i < this.length; i++) {
            if (this.array[i] == data){
                idx = i;
                break;
            }
        }

        return idx;
    }

    int findLast(int data){
        int idx = -1;
        for (int i = this.length - 1; i >= 0; i--) {
            if (this.array[i] == data){
                idx = i;
                break;
            }
        }

        return idx;
    }

    boolean exists(int data){
        return this.findFirst(data) != -1;
    }

    int getValue(int idx) {
        return this.array[idx];
    }
}

File: JavaScript/dynamic_array/dynamic_array.js
class DynamicArray {
    // Constructors
    constructor() {
        this._length = 0;
        this._capacity = 1;
        this._array = int[this._capacity];
        this._scaleFactor = 2;
    }

    constructor(scaleFactor, capacity) {
        this._length = 0;
        this._capacity = capacity;
        this._array = int[this._capacity];
        this._scaleFactor = scaleFactor;
    }

    constructor(scaleFactor, length, defaultValue) {
        this._length = length;
        this._capacity = length;
        this._array = int[this._capacity];
        this._scaleFactor = scaleFactor;

        for (let i = 0; i < this._length; i++) {
            this._array[i] = defaultValue;
        }
    }

    constructor(array) {
        this._length = array.length;
        this._capacity = array.length;
        this._array = array;
        this._scaleFactor = 2;
    }

    constructor(array, scaleFactor) {
        this._length = array.length;
        this._capacity = array.length;
        this._array = array;
        this._scaleFactor = scaleFactor;
    }

    // Public Methods
    resize() {
        this._capacity *= this._scaleFactor;
        const newArray = int[this._capacity];
        for (let i = 0; i < this._length; i++) {
            newArray[i] = this._array[i];
        }
        this._array = newArray;
    }

    append(data) {
        if (this._length == this._capacity) {
            this.resize();
        }

        this._array[this._length] = data;
        this._length++;
    }

    prepend(data) {
        if (this._length == this._capacity) {
            this.resize();
        }

        for (let i = 1; i < this._length; i++) {
            this._array[i] = this._array[i - 1];
        }

        this._array[0] = data;
        this._length++;
    }

    insert(data, idx) {
        if (idx <= 0) {
            this.prepend(data);
            return;
        }

        if (idx >= this._length) {
            this.append(data);
            return;
        }

        if (this._length == this._capacity) {
            this.resize();
        }

        for (let i = idx; i < this._array.length + 1; i++) {
            this._array[i] = this._array[i - 1];
        }

        this._array[idx] = data;
    }

    findFirst(data) {
        for (let i = 0; i < this._length; i++) {
            if (this._array[i] == data) {
                return i;
            }
        }

        return -1;
    }

    findLast(data) {
        for (let i = this._length - 1; i >= 0; i--) {
            if (this._array[i] == data) {
                return i;
            }
        }

        return -1;
    }

    exists(data) {
        return this.findFirst(data) != -1;
    }
}